### PR TITLE
[libcxx] Cache file attributes during directory iteration.

### PR DESCRIPTION
**Pull request by ed-sat**
Added caching file attributes during directory iteration for OS Windows. Allows improving performance working with files in directory.

### CODE REVIEW

**Review by EricWF — CHANGES_REQUESTED** [review]
I'm concerned this opens the library up to more TOCTOU attacks. Performance is one consideration, but correctness and security are others. Is it possible for the cached values to become stale/incorrect?

I would like to see some tests that  the cached values are actually correct when entries in the directory change.

**Comment by georgthegreat** [comment]
TL;DR: I do not consider _more TOCTOU_ and _less TOCTOU_ as the correct method of thinking. The implementation either protects you from TOCTOU (none does) or not.

@EricWF, thanks for pointing to this issue.

**There is, of course, TOCTOU issue with the code**, yet this PR neither introduced it nor do we consider it as a problem.

filesystem is a naturally shared resource not controlled by a single process, hence one should either 

* do explicit synchronization / locking it from concurrent usage (I am not aware of any implementation that would follow this option)

OR

* do nothing and be OK with slightly out of date information being returned.

**I do not think that either option is an std library responsibility.**
As the [standard prescribes](https://eel.is/c++draft/fs.class.directory.entry#general-2):

> Implementations should store such additional file attributes during directory iteration if their values are available and storing the values would allow the implementation to eliminate file system accesses by directory_entry observer functions

So, technically speaking, this PR makes current implementation more conformant to the standard than it was before.

**Now, let's observe particular consequences brought by this PR.**

First of all, this PR only affects `_LIBCPP_WIN32API` implementation.
Prior to this PR TOCTOU only affected file_type (which was cached already) and now it affects both `file_size` and last_write_time(). This does not seem as a degradation to me.

POSIX-implementations will require additional lstat / stat syscall upon accessing the first attribute. libc++ will cache attributes inside directory_entry thus making it susceptible to TOCTOU, though this will happen one syscall later than for windows implementation. Once this values get cached, no additional syscalls / attribute invalidations will happen.

In both implementations the use can call `refresh()` to request cache invalidation.

**References to other implementations**

* [microsoft/STL caches](https://github.com/microsoft/STL/blob/ff0cff1ad6de63525d0a6646b49cc10667446682/stl/inc/filesystem#L2540) attributes during iterations. Cached information is TOCTOU-affected right after receiving this information from `NextFileW` syscall.
* [libstdc++ does not cache anything](https://github.com/gcc-mirror/gcc/blob/07cdba6294756af350198fbb01ea8c8efeac54dd/libstdc%2B%2B-v3/include/bits/fs_dir.h#L268) thus causing a syscall to access any attribute. The received information becomes TOCTOU-affected right after finishing the syscall and returning information to the user.

**Comment by georgthegreat** [comment]
There is also [p0317](https://open-std.org/JTC1/SC22/WG21/docs/papers/2016/p0317r0.html) with a thorough considerations on the subject.

**Comment by EricWF** [comment]
Thank you for the detailed response. I really appreciate all of the context and careful consideration. You've thoroughly addressed my question.

I would still request that we add tests  that observe the change in behavior. At the moment, the caching behavior could regress without notice.

File: libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+using namespace fs;
+
+#if defined(_WIN32)
+static void set_last_write_time_in_iteration(const fs::path& dir)
+{
+    // Windows can postpone updating last write time for file especially for
+    // directory because last write time of directory depends of its childs.
+    // See
+    // https://learn.microsoft.com/en-us/windows/win32/sysinfo/file-times
+    // To force updating file entries calls "last_write_time" with own value.
```
Comment (ed-sat) [inline libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp:35]: @EricWF, please, take a look onto this point.
As learn.microsoft.com says, FindFirstFile / FindNextFile can give outdated mtime for directories, unless it's explicitly touched by certain side call
(e. g. GetFileInformationByHandle() or Mark Russinovich File Monitor).

Here we decided to follow microsoft/STL way of handling this issue (that is, do nothing and return / cache whatever is returned by the API).

**Review by EricWF — DISMISSED** [review]
Please wait until the bots come back.

But otherwise it looks good to me.

File: libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp (hunk 1 above)
Comment (EricWF) [reply libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp]: I appreciate the explicit documentation in the test. I think that behavior is a good choice.

**Comment by georgthegreat** [comment]
@EricWF, could you, please approve it once more?

The latest version should have the CI fixed.
Unfortunately, we are unable to reproduce buildkite/ failures locally.

Could you, please, provide us with some clues?

**Comment by ldionne** [comment]
@georgthegreat Approved the workflow just now. The CI failure appears to be happening on macOS. It's too late right now but I will have a look tomorrow or in the coming days, I should be able to help a bit since I have access to these machines more easily than most people.

**Comment by ldionne** [comment]
Ahhhh, I understand what's going on. I think that as-is, this change is actually an ABI break but there should be a way to deploy it.

Basically, `libcxx/include/__filesystem/directory_entry.h` now relies on the fact that the shared library is filling up all the right attributes within a cached entry. But if you run against an older shared library that doesn't contain the changes in this PR, the cached entries created by that shared lib will not be fully populated. This scenario can happen if you e.g. compile an application against the latest libc++ headers and then try to deploy it on an older operating system that uses a previously-compiled libc++ shared library. That's what we call the "back-deployment use case" and we typically test it on macOS, which is what the CI failures are about.

The typical way of making changes of this kind is to guard the new code you want to introduce behind a macro that basically says "I want promise I am not deploying to a system that doesn't have the changes in this PR". You can look at how e.g. `_LIBCPP_AVAILABILITY_HAS_TZDB` handles it in `__configuration/availability.h` for examples, but basically you'd need to guard your changes with something like

```c++
#if _LIBCPP_AVAILABILITY_HAS_FULLY_POPULATED_CACHED_ENTRY
...
#endif
```

where, in `__configuration/availability.h`, you define:

```c++
#define _LIBCPP_AVAILABILITY_HAS_FULLY_POPULATED_CACHED_ENTRY _LIBCPP_INTRODUCED_IN_LLVM_19
```

You might also want to slightly reorganize the code changes in your PR to avoid making it a maze of `#ifdef`s.

**Comment by georgthegreat** [comment]
@ldionne, thanks a lot. We use fully static linkage, so ABI breakages are not something that we use to work with.
We will investigate the break and provide the solution.

**Review by ldionne — CHANGES_REQUESTED** [review]
Requesting changes just so it shows up correctly in the PR queue. Let me know if you need help with implementing what I outlined above.

**Comment by ed-sat** [comment]
@ldionne 
I rewrote PR following your instructions. 
@ldionne  @ldionne Please take a look. Thank you for patience Unfortunately I was very busy on my work and didn't have time to rewrite PR straight away.

**Comment by georgthegreat** [comment]
@ldionne, can we proceed with this PR?

**Review by ldionne — COMMENTED** [review]
A few more comments but this looks really close to being good to go!

**Comment by ldionne** [comment]
Just a reminder that the LLVM 19 branch point is happening next week, so if we want this to make it into the branch we'd have to land it soon.

**Comment by ldionne** [comment]
Ping. It would be great to have this in LLVM 19. The sooner we have this in, the sooner we can eventually remove the backdeployment logic (a long time from now but still).

**Comment by ed-sat** [comment]
> Ping. It would be great to have this in LLVM 19. The sooner we have this in, the sooner we can eventually remove the backdeployment logic (a long time from now but still).

Hello. Sorry I was very busy on my work. I will spend today and weekends on fixes.

File: libcxx/src/filesystem/file_descriptor.h [hunk 2]
```diff
@@ -97,11 +97,18 @@ inline uintmax_t get_file_size(const WIN32_FIND_DATAW& data) {
...
+  auto ts = filetime_to_timespec(time);
+  if (!fs_time::is_representable(ts))
+    return file_time_type::min();
+  return fs_time::convert_from_timespec(ts);
+}
+inline perms get_file_perm(const WIN32_FIND_DATAW& data) {
+  unsigned st_mode = 0555; // Read-only
+  if (!(data.dwFileAttributes & FILE_ATTRIBUTE_READONLY))
+    st_mode |= 0222; // Write
+  return static_cast<perms>(st_mode) & perms::mask;
```
[the remark this answers is not in the document]
Comment (ed-sat) [reply libcxx/src/filesystem/file_descriptor.h:111]: `get_write_time` has never been used before and had wrong implementation and I wrote implementation based on 
https://github.com/llvm/llvm-project/blob/main/libcxx/src/filesystem/time_utils.h#L334
`__extract_last_write_time`

About `get_file_perm`, implementation was written based on code from
https://github.com/llvm/llvm-project/blob/main/libcxx/src/filesystem/posix_compat.h#L160
```
inline int stat_handle(HANDLE h, StatT* buf)
...
  buf->st_mode = 0555; // Read-only
  if (!(basic.FileAttributes & FILE_ATTRIBUTE_READONLY))
    buf->st_mode |= 0222; // Write
```

File: libcxx/include/__filesystem/directory_entry.h [hunk 3]
```diff
@@ -201,7 +201,11 @@ class directory_entry {
     _IterNonSymlink,
     _RefreshSymlink,
     _RefreshSymlinkUnresolved,
-    _RefreshNonSymlink
+    _RefreshNonSymlink,
+#  if _LIBCPP_AVAILABILITY_HAS_FILESYSTEM_FULLY_POPULATED_CACHED_ENTRY
```
[the remark this answers is not in the document]
Comment (ed-sat) [reply libcxx/include/__filesystem/directory_entry.h:205]: I have just been thinking about should we really use define `_LIBCPP_AVAILABILITY_HAS_FILESYSTEM_FULLY_POPULATED_CACHED_ENTRY`
Because If I add `_RefreshNonSymlink` and ` _RefreshNonSymlink`
   in the end of enum it will mean that the new filesystem source code in trying to use the old dylib will stay correct because the old dylib will use old enum values in filling directory_entry and enum order didnt' change.

File: libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: can-create-symlinks
+// UNSUPPORTED: c++03, c++11, c++14
+// UNSUPPORTED: no-filesystem
+// UNSUPPORTED: availability-filesystem-missing
+
+// <filesystem>
+
+// recursive_directory_iterator
```
[the remark this answers is not in the document]
Comment (ed-sat) [reply libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp:16]: I have added comment before test and wrote test for testing cached properties on Windows.

File: libcxx/src/filesystem/directory_iterator.cpp [hunk 5]
```diff
@@ -77,13 +77,13 @@ class __dir_stream {
   bool assign() {
     if (!wcscmp(__data_.cFileName, L".") || !wcscmp(__data_.cFileName, L".."))
       return false;
-    // FIXME: Cache more of this
-    // directory_entry::__cached_data cdata;
-    // cdata.__type_ = get_file_type(__data_);
-    // cdata.__size_ = get_file_size(__data_);
-    // cdata.__write_time_ = get_write_time(__data_);
     __entry_.__assign_iter_entry(
```
[the remark this answers is not in the document]
Comment (ed-sat) [reply libcxx/src/filesystem/directory_iterator.cpp:80]: I don't understand exactly your question. But for my it was easy to write a new function because 
__create_iter_result is being used in other OS implementation and I need only for Windows

File: libcxx/include/__filesystem/directory_entry.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ed-sat) [reply libcxx/include/__filesystem/directory_entry.h]: I had thoughts about reusing `_IterSymlink` and `_IterNonSymlink` in my first attempt to add caching and I faced that I would be kinda difficult because in that case I should store information about cached or not properties depending on platform Windows and others. And also based on that knowledge rewrite many functions with enum switch by adding for each case with  `_IterSymlink` and `_IterNonSymlink` additional login which quite similar to  '_IterCachedSymlink' and '_IterCachedNonSymlink'
In that case each plain enum switch will become enum with additional if else.

[the remark this answers is not in the document]
Comment (ed-sat) [reply libcxx/include/__filesystem/directory_entry.h]: I have prepared patch without using availability define. If it suits for you we manage to release this patch with LLVM 19. Otherwise I can try to rework this patch like in the first option and release it later.

**Comment by ed-sat** [comment]
Hello @ldionne
Do we have opportunity to merge this patch or I need to rework it?  For this patch dashboard shows that all test passed.

**Comment by georgthegreat** [comment]
@ldionne, it looks like the current implementation is both ABI-stable and straight forward to review.
Maybe we can proceed with merging it?

**Comment by georgthegreat** [comment]
@ldionne, gentle ping.

**Comment by georgthegreat** [comment]
@ldionne, let's merge it!

**Review by ldionne — APPROVED** [review]
LGTM with minor comments.

File: libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+#include <filesystem>
+#include <type_traits>
+#include <set>
+#include <cassert>
+
+#include "test_macros.h"
+#include "filesystem_test_helper.h"
+namespace fs = std::filesystem;
+using namespace fs;
```
[the remark this answers is not in the document]
Comment (ed-sat) [reply libcxx/test/std/input.output/filesystems/class.rec.dir.itr/cache_refresh_iter.pass.cpp]: I removed `using namespace fs;`

**Comment by ed-sat** [comment]
@EricWF Please take a look after all changes and fixes.

**Review by ldionne — APPROVED** [review]
LGTM once CI is green. The ARM CI might be failing due to the zdump issue -- if so that's unrelated to this patch and shouldn't block merging. Whoever sees the green CI first can merge.

**Comment by georgthegreat** [comment]
@ldionne, sorry, but could you help us to interpret the failures? It looks like aix tests have failed too...
